### ISSUE TITLE
How to accomplish no-vendor-prefix rules

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#103]
I see two routes (maybe there are more):

- The "simple" route: Simply disallow all vendor prefixes. If a person wants to use a non-standard property with a vendor-prefix (e.g. `-webkit-appearance`) they will need to *turn this rule off* and then *turn it back on* ---- so this route is only feasible if we have inline-disable-and-enable implemented.
- The "complex" route: Use a list to determine which rules should *not* get prefixes. This way, the linter will not complain if you use `-webkit-appearance`: it won't be on the list of prefixed-properties to watch out for.

How to do the complex route? I'd say we could manually maintain a list based on what Autoprefixer looks for; or we could automatically use Autoprefixer's data by adding Autoprefixer as a dependency. It looks to me like we could get an array of properties and at-rules that will receive vendor prefixes from Autoprefixer by checking `Object.keys(require('autoprefixer-core').data.prefixes)`. @ai maybe you can confirm?

I originally made this rule for SCSS Lint here: https://github.com/brigade/scss-lint/pull/250. There we opted for the "complex" route. I think that inline-disabling and -enabling was not built in and there wasn't much indication that it would be any time soon, so I didn't think of that as a valid route at the time.

Thoughts?

**Comment by MoOx** [comment]
I saw another possible way (a bit different as the complexe one, but maybe we can offer this one too): only check prefixed value (-(webkit|moz|...)-* and read siblings to yell only if there is already the non prefixed value too. This way if someone use a prefix by hand, you can warn him he do not need to do that. This also allow usage when there is no non-prefixed equivalent... But well maybe it's a good solution...

You complex seems the better, if coupled with a nice up to date database. If we can easily rely on Autoprefixer that would be awesome !

**Comment by ai** [comment]
@davidtheclark is a variant for check:

1. Run Autoprefixer on CSS with `browsers: []`
2. Diff CSS before/after Autoprefixer
3. Warn on every removed property.

Maybe I can add some API if you suggest some simle solution.

**Comment by davidtheclark** [comment]
Latest effort is here: d76e53344e2d0748c8cb3d5b8e6903bb1a7db603 (the `declaration-no-vendor-prefix` branch).

How I'm doing it is this:
- For each rule ...
- Check if any prop or value has a prefix; if not, ignore
- Run the rule through Autoprefixer with `{ add: false, browsers: [] }`
- Check whether all props and values from original rule are present in the autoprefixed rule
- Any props/values not present in the autoprefixed rule trigger a warning

Here are the limitations I know of:
- It will only flag vendor prefixes that Autoprefixer would change. Anything Autoprefixer would not change is ignored. So this method will *not* flag a couple things it might be nice to flag:
  - mistaken prefixes, like `-o-columns`
  - prefixed identifiers occurring after non-prefixed identifiers, like `column: 2; -webkit-columns: 2;`

But maybe that's the best we can do. In order to identify and flag mistaken prefixes, we would have to have data telling us all of the prefixed identifiers that are not actually useful -- that's probably unreasonable. (If somebody puts `-foo-bar: 2`, the nature of CSS is that that useless declaration will just be ignored by the browser.) And we might want to allow the second issue (prefixed *after* non-prefixed) for the same reason Autoprefixer does: to allow people to hack if they need to.

Thoughts?

**Comment by davidtheclark** [comment]
A possible alternative method with different downsides:

- For each rule ...
- Check if any prop or value has a prefix; if not, ignore
- Strip the prefix using https://github.com/postcss/postcss/blob/master/docs/api.md#vendorunprefixedstring
- See if the stripped identifier is included in Autoprefixer's data-list
- If it is, then the prefix was unnecessary so trigger a warning

The advantages:
- It will catch mistaken prefixes
- It won't care whether prefixes occur before or after the non-prefixed identifier
- We don't have to run  every occurrence through Autoprefixer, so it will probably be faster

Disadvantages:
- It will not know how to warn for "hacks" (e.g. old flexbox syntax, where the identifier stripped of its prefix does not correspond to an item in Autoprefixer's data list)
- It will not allow you to use vendor prefixes in hacky ways except by turning off the rule

**Comment by MoOx** [comment]
What about something simpler and stupid for a start ?
If someone use a prefixed rule that is already being unprefixed with the same (-prefix) value -> warning.

```css
/* warning */
sel {
  -webkit-stuff: -webkit-thing;
  stuff: thing;
}


/* no warning */
sel {
  -webkit-stuff: -webkit-whatever;
  stuff: different value;
}
```

I think this will cover most usages like when new people on a project add a -*- + unprefixed ?

We should make a simple rule otherwise we are going to get lot's of issue opened for "not that much" edge cases.

**Comment by ai** [comment]
@davidtheclark Autoprefixer already contains data of common mistaken prefixes https://github.com/postcss/autoprefixer-core/blob/master/data/prefixes.coffee#L51

**Comment by ai** [comment]
BTW, here is a code to convert `-webkit-X` to `Y`: https://github.com/postcss/autoprefixer-core/blob/master/lib/prefixes.coffee

**Comment by davidtheclark** [comment]
Thank you @ai! I will look into that code.

**Comment by davidtheclark** [comment]
Made another attempt: 7b0b654848be07415b713968cd5f7f7bad75a29b

@ai You can see I tried to use the code you referenced in these lines:

```js
const prefixes = new Prefixes(
  autoprefixer.data.prefixes,
  new Browsers(autoprefixer.data.browsers, [])
)
```

So I am drawing on the non-public API. I don't suppose there's a way I'm missing to access this functionality through the public API?

Also, the `unprefixed()` function generally works, but I haven't gotten it to work for hack values, such as `display: -webkit-box` (it unprefixes to `box` instead of `flex`) or `image-rendering: -webkit-optimize-contrast` (unprefixes to `optimize-contrast` instead of `crisp-edges`). Am I missing something, or is this inevitable?

@MoOx I'm ok with making a simpler version, too, of course. I thought I may as well give it a few shots first, though :)

**Comment by ai** [comment]
@davidtheclark prefixed values is mucch more difficult (it can be mixed with other values like in `background`), so Autoprefixer can’t overt it back to unprefixed.

**Comment by ai** [comment]
@davidtheclark yeap, `Prefixes` is a non-public API so you use right way to create instance.

**Comment by davidtheclark** [comment]
Thanks for taking a look @ai.

Seems, then, like this approach might be the best we can do -- and it's really very good ... the only thing it stumbles on is values. @MoOx do you agree this would be a good way to proceed?

**Comment by MoOx** [comment]
lgtm

**Comment by davidtheclark** [comment]
I think this went pretty well for the `property` rule. Just need to implement for `value` and `selector` and `media`.
